React API (#16)
reactjs/fr.reactjs.org: content/docs/reference-react.md

---
id: react-api

> title: L'API haut-niveau de React

Reviewer: En fait pour la navigation on reste sur le namespace, comme pour ReactDOM, SyntheticEvent, etc. 

J'ai récemment mis a jour toute la navigation dans master, c'est elle qui fait référence. Ça n'empêche pas d'avoir des titres plus explicites au sein de la page.

> title: L'API haut-niveau de React

Author: Ok :)

)
```

Cette méthode clone et renvoie un nouvel élément en utilisant `element` comme point de départ. L'élément obtenu aura les props de l'élément originel sur lesquelles ont aura effectué une fusion de surface des nouvelles props. Les nouveaux éléments enfants (`children`) replaceront les anciens. les `key` et `ref` issues de l'élément originel seront préservées.

`React.cloneElement()` est quasiment équivalent à :

```js
<element.type {...element.props} {...props}>{children}</element.type>
```

> Cependant elle préserve les `ref`. Concrètement, ça signifie que si vous avez un élément enfant avec une `ref` qui lui est attaché, vous ne le volerez pas accidentellement à son ancêtre. Vous aurez la même `ref` attaché à un nouvel élément.

Author: Plutôt crever que de pluraliser `ref` dans ce contexte. Le fait que l'on pluralise un fragment de code n'a aucun sens en français... y'a bien que les anglais pour transformer tout et n'importe quoi en nom dénombrable !

> Cependant elle préserve les `ref`. Concrètement, ça signifie que si vous avez un élément enfant avec une `ref` qui lui est attaché, vous ne le volerez pas accidentellement à son ancêtre. Vous aurez la même `ref` attaché à un nouvel élément.

Reviewer: Ok mais tu prends le reste de la suggestion ?

React.isValidElement(object)
```

Cette méthode vérifie qu'un objet est bien un élément React. Elle renvoie `true` ou `false`.

* * *

### `React.Children` {#reactchildren}

`React.Children` fournit des utilitaires pour interagir avec la structure de données opaque de `this.props.children`.

#### `React.Children.map` {#reactchildrenmap}

```javascript
React.Children.map(children, function[(thisArg)])
```

> Cette méthode exécute une fonction sur chacun des enfants direct contenu dans `children` avec `this` défini pour être `thisArg`. Si `children` est un tableau, il sera parcouru et la fonction sera appelée sur chacun des enfants contenu dans ce tableau. Si `children` est `null` ou `undefined`, elle renverra `null` ou `undefined` plutôt qu'un tableau.

Author: J'avoue que cette explication m'a bien dérouté, je suis tout a fait d'accord pour suggérer un changement dans la VO !

#### `React.Children.count` {#reactchildrencount}

```javascript
React.Children.count(children)
```

Cette méthode renvoie le nombre total de composant présent dans `children`, ce total étant égale au nombre de fois ou une fonction de rappel passée à `map` or `forEach` serait exécutée.

#### `React.Children.only` {#reactchildrenonly}

```javascript
React.Children.only(children)
```

Cette méthode vérifie que `children` n'a qu'un seul enfant (un élément React) et le renvoie. Si ce n'est pas le cas elle lancera une erreur.

> Remarque :
>
> `React.Children.only()` n'accepte pas le type de valeur retourné par [`React.Children.map()`](#reactchildrenmap) car il s'agit d'un tableaux et pas d'un éléments React.

#### `React.Children.toArray` {#reactchildrentoarray}

```javascript
React.Children.toArray(children)
```

Cette méthode renvoie la structure de donnée opaque de `children` sous la forme d'un tableau linéarisé ou chaque enfant se voie assigné une `key`. C'est utile si vous voulez manipuler une collection d'enfants dans votre méthode de rendu, en particulier si vous voulez réorganiser ou réduire `this.props.children` avant de le passer à d'autres éléments.

> Remarque :
>

> > `React.Children.toArray()` change les `key` pour préserver la sémantique des tableaux imbriqués pendant la linéarisation des enfants. Sa signifie que `toArray` préfixe chaque `key` dans le tableau qui sera renvoyé de manière a ce que la `key` de chaque élément soit associée au tableau originel qui les contient.

Author: 100% d'accord avec toi !

### `React.forwardRef` {#reactforwardref}

`React.forwardRef` créée un composant React qui transfert la valeur de l'attribut [ref](/docs/refs-and-the-dom.html) qu'il reçoit à un autre composant plus bas dans l'arbre. Cette technique est assez inhabituel mais elle est particulièrement utile dans deux cas :

* [Transférer une référence à un composant DOM](/docs/forwarding-refs.html#forwarding-refs-to-dom-components)
* [Transférer une référence à un composant de haut niveau](/docs/forwarding-refs.html#forwarding-refs-in-higher-order-components)

`React.forwardRef` prend une fonction de rendu comme argument. React appellera cette fonction avec deux arguments `props` et `ref`. Cette fonction devrait renvoyer un nœud React.

`embed:reference-react-forward-ref.js`

Dans l'exemple ci-avant, la fonction de rendu passé à `React.forwardRef` va recevoir comme deuxième argument la `ref` initialement donnée à l'élément `<FancyButton ref={ref}>`. Cette fonction va alors transférer la `ref` à l'élément `<button ref={ref}>`.

En conséquence, après que React ait attaché la ref, `ref.current` pointera directement vers l'instance de l'élément DOM de `<button>`.

> Pour plus d'information, lisez [Transférer les refs](/docs/forwarding-refs.html)

Reviewer: ```suggestion
Pour en savoir plus, lisez [Transférer les refs](/docs/forwarding-refs.html)
```

### `React.lazy` {#reactlazy}

`React.lazy()` vous permet de définir un composant qui sera chargé dynamiquement. Cela aide à réduire la taille du fichier initial en reportant à plus tard le chargement des composants inutiles lors du rendu initial.

> Vous pouvez apprendre comment l'utiliser en lisant [la documentation sur l'organisation du code](/docs/code-splitting.html#reactlazy). Vous pouvez aussi vouloir lire [cette article](https://medium.com/@pomber/lazy-loading-and-preloading-components-in-react-16-6-804de091c82d) qui explique comment l'utiliser en détail.

Reviewer: ```suggestion
Vous pouvez apprendre comment l'utiliser en lisant [la documentation sur la découpe du code](/docs/code-splitting.html#reactlazy). Vous voudrez peut-être aussi jeter un œil à [cet article](https://medium.com/@pomber/lazy-loading-and-preloading-components-in-react-16-6-804de091c82d), qui explique comment l'utiliser en détail.
```

- “Code splitting” a un sens global bien précis ; la VF n'est pas encore bien établie, mais « organisation » est beaucoup trop générique. Dans l'univers Webpack, qui fait référence sur le sujet, on parle en général de « découpe », donc hop.

Author: Ouais, j’étais pas sur là, merci pour l'info.

```js
// Ce composant est chargé dynamiquement
const SomeComponent = React.lazy(() => import('./SomeComponent'));
```

> Vous noterez que l'affichage d'un composant `lazy` à besoin d'un composant `<React.Suspense>` plus haut dans l'arbre de rendu. C'est de cette manière que vous pouvez spécifier un indicateur de chargement.

Reviewer: ```suggestion
Notez bien que l'affichage d'un composant `lazy` a besoin d'un composant `<React.Suspense>` plus haut dans l'arbre de rendu. C'est de cette manière que vous pouvez spécifier un indicateur de chargement.
```

> > **Remarque**

Reviewer: ```suggestion
> **Remarque :**
```

Tu as bien homogénéisé à coup de deux-points partout, on va garder ça hein ? 😉

>

> > Utiliser `React.lazy` avec un import dynamique requiers que l'environnement JS supporte les _Promises_. Vous aurez donc besoin de _polyfill_ pour IE11 et inférieur.

Reviewer: ```suggestion
> Utiliser `React.lazy` avec un `import()` dynamique requiert une prise en charge des *Promises* par  l'environnement JS. Vous aurez donc besoin d’un _polyfill_ pour IE11 et inférieurs.
```

### `React.Suspense` {#reactsuspense}

> `React.Suspense` vous permet de définir un indicateur de chargement dans le cas ou certain composants de l'arbre des composants ne sont pas près à être affiché. Aujourd'hui le **seul** cas d'usage supporté par `<React.Suspense>` est le chargement de composant différé via [`React.lazy`](#reactlazy) :

Reviewer: ```suggestion
`React.Suspense` vous permet de définir un indicateur de chargement pour le cas où certains composants plus bas dans l’arbre de rendu ne seraient pas encore prêts à être affichés. Pour le moment le **seul** cas d'usage pris en charge par `<React.Suspense>`, c’est le chargement différé de composants via [`React.lazy`](#reactlazy) :
```

```js
// Ce composant est chargé dynamiquement
const OtherComponent = React.lazy(() => import('./OtherComponent'));

function MyComponent() {
  return (

>     // Affiche <Spinner> jusqu'a ce que <OtherComponent> soit chargé

Reviewer: ```suggestion
    // Affiche <Spinner> jusqu'à ce que <OtherComponent> soit chargé
```

}
```

> Tout est détaillé dans [la documentation sur l'organisation du code](/docs/code-splitting.html#reactlazy). Vous noterez que les composants `lazy` peuvent êtres profondément enfouis dans l'arbre des enfants de `Suspense` -- Ils n'ont pas besoin d'être enveloppés individuellement. La bonne pratique consiste à placer un `<Suspense>` partout où vous souhaitez voir un indicateur de chargement et à utiliser `lazy()` partout ou vous voulez découper votre code.

Reviewer: ```suggestion
Tout ça est détaillé dans [notre guide sur la découpe du code](/docs/code-splitting.html#reactlazy). Remarquez que les composants `lazy` peuvent être profondément enfouis dans l'arbre des descendants de `Suspense`—ils n'ont pas besoin d'être enveloppés individuellement. La bonne pratique consiste à placer un `<Suspense>` aux endroits où vous souhaitez voir un indicateur de chargement, et à utiliser `lazy()` partout ou vous voulez découper votre code.
```

Bien que ce ne soit pas le cas pour le moment, nous prévoyons d'étendre les capacités de `Suspense` pour qu'il puisse gérer d'autre scénarios tel que le chargement de données. Vous pourrez en savoir plus en jetant un coup d'œil à [notre feuille de route](/blog/2018/11/27/react-16-roadmap.html).

> Remarque :
>

> > `React.lazy()` et `<React.Suspense>` ne sont pas encore supportés par `ReactDOMServer`. C'est une limitation connue qui devrait être résolu dans le futur.

Reviewer: ```suggestion
> `React.lazy()` et `<React.Suspense>` ne sont pas encore pris en charge par `ReactDOMServer`. C'est une limitation connue qui devrait être résolue à l’avenir.
```
